mirror/mirror.go

```diff
@@ -0,0 +1,371 @@
+			} else {
+				sel = selectors.adsWithStopAt(selector.RecursionLimitNone(), cidlink.Link{Cid: mc})
+			}
+
+			_, err = m.sub.Sync(ctx, m.source.ID, cid.Undef, sel, m.source.Addrs[0],
+				legs.ScopedBlockHook(func(id peer.ID, c cid.Cid, actions legs.SegmentSyncActions) {
+					// TODO: set actions next segment link to ad previous id if it is present. For
+					//      now segmentation is disabled.
+					//       Here we could be encountering HAMT or Entry Chunk so picking the next
+					//       CID is not trivial; we probably should not use segmentation for HAMT
+					//       at all.
+
+					// Prepend to the list since the mirroring should start from the oldest ad first.
+					syncedAdCids = append([]cid.Cid{c}, syncedAdCids...)
+				}),
+				// Disable segmentation until the actions in hook are handled appropriately
+				legs.ScopedSegmentDepthLimit(-1),
+			)
+
+			if err != nil {
+				log.Errorw("Failed to sync source", "err", err)
+				continue
+			}
+
+			for _, adCid := range syncedAdCids {
+				err := m.mirror(ctx, adCid)
+				if err != nil {
+					log.Errorw("Failed to mirror ad", "cid", adCid, "err", err)
+					// TODO add an option on what to do if the mirroring of an ad failed?
+				}
+			}
+
+			syncedCount := len(syncedAdCids)
+			if syncedCount > 0 {
+				latestOriginal := syncedAdCids[syncedCount-1]
+				err = m.setLatestOriginalAdCid(ctx, latestOriginal)
+				if err != nil {
+					log.Errorw("Failed to store latest original ad cid", "cid", latestOriginal, "err", err)
+				}
+			}
+		}
+	}()
+
+	return nil
+}
+
+func (m *Mirror) Shutdown() error {
+	if m.cancel != nil {
+		m.cancel()
+	}
+	return nil
+}
+
+func (m *Mirror) mirror(ctx context.Context, adCid cid.Cid) error {
+	ad, err := m.loadAd(ctx, adCid)
+	if err != nil {
+		return err
+	}
+	if err := ad.Validate(); err != nil {
+		return err
+	}
+	// Set previous ID
+	prevMirroredAdCid, err := m.getLatestMirroredAdCid(ctx)
+	if err != nil {
+		return err
+	}
+	switch prevMirroredAdCid {
+	case cid.Undef:
+		ad.PreviousID = nil
+	default:
+		ad.PreviousID = cidlink.Link{Cid: prevMirroredAdCid}
+	}
+	entriesCid := ad.Entries.(cidlink.Link).Cid
+	if !ad.IsRm {
+		switch entriesCid {
+		case cid.Undef:
+			// advertisement is invalid? entries CID should never be cid.Undef for non-removal ads.
+			return errors.New("entries link is cid.Undef")
+		case schema.NoEntries.Cid:
+			// Nothing to do.
+		default:
+			// TODO: it is unfortunate that go-legs takes a single address here... this needs to be
+			//       fixed in go-legs.
+			_, err = m.sub.Sync(ctx, m.source.ID, entriesCid, selectors.entriesWithLimit(m.entriesRecurLimit), m.source.Addrs[0])
+			if err != nil {
+				log.Errorw("Failed to sync entries", "cid", entriesCid, "err", err)
+				return err
+			}
+			ad.Entries, err = m.remapEntries(ctx, ad.Entries)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	if err := ad.Sign(m.h.Peerstore().PrivKey(m.h.ID())); err != nil {
```

Comment:
if we don't re-map, we could i guess retain the original signature?

Reply:
Sure. I am curious what's the benefit in not resigning the ad? I can imagine a scenario where a mirror with persistent state is reconfigured such that the ad signature would differ. This would then mean some ads in the chain are signed by the original provider and some ads are signed by the mirror. 
Aside from that we need logic to determine when to resign and when not to resign. 

So I wonder is it not simpler to just resign ads by the mirror? and what's the benefit of not resigning considering a mirror always acts as a CDN for the original ad chain anyway?

Reply:
Changed the default behaviour to preserve the original signature whenever possible by default, with an option to always re-sign ads with the mirror's identity.

A mirror configured to not re-map the entries, would then mirror the original ad chain identically without any change. Tests are added to assert this. 

Please shout if there is anything else I might have missed.